File: js/src/models/anthropic.ts
// Anthropic §5 HTTP model client — grounded answers over the native Messages
// API (SPEC-PORTS-v1). Anthropic is not OpenAI-shaped, so it gets its own client:
// POST {base}/v1/messages, `system` as a TOP-LEVEL field, a REQUIRED max_tokens,
// and a content[] response whose text blocks are concatenated in order.
//
// Like every §5 client the HTTP call goes through an injected transport, headers
// are ALWAYS just {"Content-Type":"application/json"} (auth is the endpoint
// layer's job), and temperature is always sent (default 0.0 — deterministic).

import { wireHeaders } from "../http.js";
import { SYSTEM_PROMPT, userMessage, type Transport } from "./openai.js";

/** Anthropic requires max_tokens; use a sane default when the caller gives none. */
const DEFAULT_MAX_TOKENS = 1024;

export interface AnthropicConfig {
  /** Defaults to Anthropic's public API base. */
  base_url?: string;
  model: string;
  temperature?: number;
  /** Required by the API; defaults to 1024. */
  max_tokens?: number;
  /** First-class auth/provider headers as ${ENV} templates, e.g.
   *  `{ "x-api-key": "${ANTHROPIC_API_KEY}" }` — expanded at call time. */
  headers?: Record<string, string>;
}

/** Grounded answers over Anthropic's native `/v1/messages` endpoint. */
export class AnthropicGenerator {
  private readonly baseUrl: string;
  private readonly model: string;
  private readonly temperature: number;
  private readonly maxTokens: number;
  private readonly transport: Transport;
  private readonly headers: Record<string, string> | undefined;

  constructor(config: AnthropicConfig, transport: Transport) {
    this.baseUrl = (config.base_url ?? "https://api.anthropic.com").replace(/\/+$/, "");
    this.model = config.model;
    this.temperature = config.temperature ?? 0.0;
    this.maxTokens = config.max_tokens ?? DEFAULT_MAX_TOKENS;
    this.transport = transport;
    this.headers = config.headers;
  }

  answer(question: string, passage: string, answerLanguage = "en"): string {
    const request: Record<string, unknown> = {
      model: this.model,
      system: SYSTEM_PROMPT,
      messages: [
        { role: "user", content: userMessage(question, passage, answerLanguage) },
      ],
      // max_tokens is required; temperature keeps answers deterministic (§5).
      max_tokens: this.maxTokens,
      temperature: this.temperature,
    };
    const raw = this.transport(
      `${this.baseUrl}/v1/messages`,
      JSON.stringify(request),
      wireHeaders(this.headers),
    );
    const payload = JSON.parse(raw) as {
      content?: { type?: string; text?: string }[];
    };
    const blocks = payload.content ?? [];
    return blocks
      .filter((block) => block.type === "text")
      .map((block) => block.text ?? "")
      .join("");
  }
}

File: golang/models/anthropic.go
package models

import (
	"encoding/json"
	"strings"
)

// DefaultAnthropicBaseURL is used when the caller passes an empty base URL.
const DefaultAnthropicBaseURL = "https://api.anthropic.com"

// DefaultAnthropicMaxTokens is Anthropic's required max_tokens default (§4b).
const DefaultAnthropicMaxTokens = 1024

// AnthropicGenerator posts grounded answers to Anthropic's native /v1/messages
// endpoint (§4b): system is a top-level field, max_tokens is REQUIRED, and the
// reply is the concatenation of the text blocks in content[].
type AnthropicGenerator struct {
	baseURL     string
	model       string
	temperature float64
	maxTokens   int
	transport   Transport
	headers     map[string]string
}

// NewAnthropicGenerator builds an Anthropic generator. An empty baseURL defaults
// to https://api.anthropic.com; trailing "/" is stripped. Pass WithHeaders(...)
// for first-class ${ENV} auth headers (e.g. {"x-api-key": "${ANTHROPIC_API_KEY}"}).
func NewAnthropicGenerator(baseURL, model string, temperature float64, maxTokens int, transport Transport, opts ...Option) *AnthropicGenerator {
	if baseURL == "" {
		baseURL = DefaultAnthropicBaseURL
	}
	return &AnthropicGenerator{
		baseURL:     strings.TrimRight(baseURL, "/"),
		model:       model,
		temperature: temperature,
		maxTokens:   maxTokens,
		transport:   transport,
		headers:     applyOptions(opts),
	}
}

func (g *AnthropicGenerator) endpoint() string {
	return g.baseURL + "/v1/messages"
}

// Answer generates a grounded answer from passage in answerLanguage.
func (g *AnthropicGenerator) Answer(question, passage, answerLanguage string) (string, error) {
	request := map[string]any{
		"model":  g.model,
		"system": SystemPrompt,
		"messages": []any{
			map[string]any{"role": "user", "content": buildUserPrompt(question, passage, answerLanguage)},
		},
		// max_tokens is required; temperature keeps answers deterministic (§4b).
		"max_tokens":  g.maxTokens,
		"temperature": g.temperature,
	}
	body, err := json.Marshal(request)
	if err != nil {
		return "", err
	}
	raw, err := g.transport(g.endpoint(), body, wireHeaders(g.headers))
	if err != nil {
		return "", err
	}
	var payload struct {
		Content []struct {
			Type string `json:"type"`
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.Unmarshal(raw, &payload); err != nil {
		return "", err
	}
	var sb strings.Builder
	for _, block := range payload.Content {
		if block.Type == "text" {
			sb.WriteString(block.Text)
		}
	}
	return sb.String(), nil
}
